sklearn/impute/tests/test_knn.py

```diff
@@ -0,0 +1,638 @@
+        [1, 0, 0, 1],
+        [0, na, 1, na],
+        [1, 1, 1, na],
+        [0, 1, 0, 0],
+        [0, 0, 0, 0],
+        [1, 0, 1, 1],
+        [10, 10, 10, 10],
+    ])
+
+    # Get weights of donor neighbors
+    dist = nan_euclidean_distances(X)
+    r1c1_nbor_dists = dist[1, [0, 2, 3, 4, 5]]
+    r1c3_nbor_dists = dist[1, [0, 3, 4, 5, 6]]
+    r1c1_nbor_wt = (1 / r1c1_nbor_dists)
+    r1c3_nbor_wt = (1 / r1c3_nbor_dists)
+
+    r2c3_nbor_dists = dist[2, [0, 3, 4, 5, 6]]
+    r2c3_nbor_wt = 1 / r2c3_nbor_dists
+
+    # Collect donor values
+    col1_donor_values = np.ma.masked_invalid(X[[0, 2, 3, 4, 5], 1]).copy()
+    col3_donor_values = np.ma.masked_invalid(X[[0, 3, 4, 5, 6], 3]).copy()
+
+    # Final imputed values
+    r1c1_imp = np.ma.average(col1_donor_values, weights=r1c1_nbor_wt)
+    r1c3_imp = np.ma.average(col3_donor_values, weights=r1c3_nbor_wt)
+    r2c3_imp = np.ma.average(col3_donor_values, weights=r2c3_nbor_wt)
+
+    X_imputed = np.array([
+        [1, 0, 0, 1],
+        [0, r1c1_imp, 1, r1c3_imp],
+        [1, 1, 1, r2c3_imp],
+        [0, 1, 0, 0],
+        [0, 0, 0, 0],
+        [1, 0, 1, 1],
+        [10, 10, 10, 10],
+    ])
+
+    imputer = KNNImputer(weights="distance")
+    assert_allclose(imputer.fit_transform(X), X_imputed)
+
+
+def test_knn_imputer_callable_metric():
+
+    # Define callable metric that returns the l1 norm:
+    def custom_callable(x, y, missing_values=np.nan, squared=False):
+        x = np.ma.array(x, mask=np.isnan(x))
+        y = np.ma.array(y, mask=np.isnan(y))
+        dist = np.nansum(np.abs(x-y))
+        return dist
+
+    X = np.array([
+        [4, 3, 3, np.nan],
+        [6, 9, 6, 9],
+        [4, 8, 6, 9],
+        [np.nan, 9, 11, 10.]
+    ])
+
+    X_0_3 = (9 + 9) / 2
+    X_3_0 = (6 + 4) / 2
+    X_imputed = np.array([
+        [4, 3, 3, X_0_3],
+        [6, 9, 6, 9],
+        [4, 8, 6, 9],
+        [X_3_0, 9, 11, 10.]
+    ])
+
+    imputer = KNNImputer(n_neighbors=2, metric=custom_callable)
+    assert_allclose(imputer.fit_transform(X), X_imputed)
+
+
+@pytest.mark.parametrize("na", [-1, np.nan])
+def test_knn_imputer_with_simple_example(na):
+
+    X = np.array([
+        [0, na, 0, na],
+        [1, 1, 1, na],
+        [2, 2, na,  2],
+        [3, 3, 3, 3],
+        [4, 4, 4, 4],
+        [5, 5, 5, 5],
+        [6, 6, 6, 6],
+        [na, 7, 7, 7]
+    ])
+
+    r0c1 = np.mean(X[1:6, 1])
+    r0c3 = np.mean(X[2:-1, -1])
+    r1c3 = np.mean(X[2:-1, -1])
+    r2c2 = np.mean(X[[0, 1, 3, 4, 5], 2])
+    r7c0 = np.mean(X[2:-1, 0])
+
+    X_imputed = np.array([
+        [0, r0c1, 0, r0c3],
+        [1, 1, 1, r1c3],
+        [2, 2, r2c2, 2],
+        [3, 3, 3, 3],
+        [4, 4, 4, 4],
+        [5, 5, 5, 5],
+        [6, 6, 6, 6],
+        [r7c0, 7, 7, 7]
+    ])
+
+    imputer_comp = KNNImputer(missing_values=na)
+    assert_allclose(imputer_comp.fit_transform(X), X_imputed)
+
+
+@pytest.mark.parametrize("na", [-1, np.nan])
+def test_knn_imputer_with_weighted_features(na):
```

Comment:
"weighted features" seems to be a strange name. This also appears to overlap in purpose with `test_knn_imputer_weight_distance`. What's the distinction?

Reply:
There was not a distinction. `test_knn_imputer_with_weighted_features` was removed and the tests from there was moved into `test_knn_imputer_weight_distance`